### PR TITLE
Create IMarkers for BUG,NOTE,FIXME,HACK,TODO,XXX comments

### PR DESCRIPTION
**Pull request by sebthom**
This PR addresses #405.

Example:
markers

The markers are asynchronously updated via:
 `TMModel.TokenizerThread#revalidateTokens()` 
&rarr; `TMPresentationReconciler.InternalListener#modelTokensChanged()` 
&rarr; `MarkerUtils.updateTextMarkers()`

### CODE REVIEW

**Comment by sebthom** [comment]
@mickaelistria @akurtakov @angelozerr please give this a try if you have a chance

File: org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/ResourceUtils.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import org.eclipse.core.filebuffers.ITextFileBufferManager;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jface.text.IDocument;
+
+public final class ResourceUtils {
+
+	@Nullable
```
Comment (angelozerr) [inline org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/ResourceUtils.java:21]: Define a private constructor

File: org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import org.eclipse.tm4e.core.model.ITMModel;
+import org.eclipse.tm4e.core.model.ModelTokensChangedEvent;
+import org.eclipse.tm4e.core.model.TMToken;
+import org.eclipse.tm4e.ui.TMUIPlugin;
+import org.eclipse.tm4e.ui.internal.model.TMDocumentModel;
+
+public final class MarkerUtils {
+
+	private abstract static class AbstractMarkerConfig {
+		final String type;
```
Comment (angelozerr) [inline org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:31]: Define a private constructor

File: org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+			final var lineNumberAttr = marker.getAttribute(IMarker.LINE_NUMBER);
+			if (lineNumberAttr instanceof final Integer lineNumber)
+				return lineNumber;
+		} catch (final CoreException ex) {
+			TMUIPlugin.log(new Status(IStatus.ERROR, TMUIPlugin.PLUGIN_ID, ex.getMessage(), ex));
+		}
+		return -1;
+	}
+
+	public static void updateTextMarkers(final ModelTokensChangedEvent event) {
```
Comment (angelozerr) [inline org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:82]: Move this staic public method before privatif methods. Add j1vadoc for this method

Comment (sebthom) [reply org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:82]: Going through the code base I actually haven't yet seen any continuous convention how methods are ordered, why do you want to have the method order changed here? Is this a best practice/convention that is documented anywhere?

Comment (angelozerr) [reply org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:86]: IMHO public code is more important than private code. When I open a Java file, I like to see on the top the public method that I can use it although private code is less important.

File: org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		}
+	}
+
+	/**
+	 * Updates all TM4E text markers of the {@link TMDocumentModel}'s document starting from the given line number until
+	 * the end of the document.
+	 *
+	 * @param startLineNumber 1-based
+	 */
+	public static void updateTextMarkers(final TMDocumentModel docModel, final int startLineNumber)
```
Comment (angelozerr) [inline org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:103]: I think you should update some existing marker and not delete all and recreate. I suggest to see the update marker of LSP4E at https://github.com/eclipse/lsp4e/blob/ca172adc3afec075357986c6c8c5459335e0e98a/org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/diagnostics/LSPDiagnosticsToMarkers.java#L154

Comment (sebthom) [reply org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:98]: I now implementing a diffing mechanism.

File: org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+import org.eclipse.tm4e.core.model.ITMModel;
+import org.eclipse.tm4e.core.model.ModelTokensChangedEvent;
+import org.eclipse.tm4e.core.model.TMToken;
+import org.eclipse.tm4e.ui.TMUIPlugin;
+import org.eclipse.tm4e.ui.internal.model.TMDocumentModel;
+
+public final class MarkerUtils {
+
+	private abstract static class AbstractMarkerConfig {
+		final String type;
```
Comment (mickaelistria) [inline org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:37]: Couldn't it be a record?

File: org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+		final int severity;
+
+		AbstractMarkerConfig(final String type, final int priority, final int severity) {
+			this.type = type;
+			this.priority = priority;
+			this.severity = severity;
+		}
+	}
+
+	private static final class ProblemMarkerConfig extends AbstractMarkerConfig {
```
Comment (mickaelistria) [inline org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java:48]: I don't like the idea of having subclasses just to specialize constructors. The parent could instead provide some `newProblemMarkerConfig(priority)` and `newTaskMarkerConfig(priorty)` methods.

**Comment by mickaelistria** [comment]
Thanks!

**Pull request merged**